rust/kernel/spi.rs

```diff
@@ -0,0 +1,180 @@
+    }
+
+    pub fn to_ptr(&mut self) -> *mut bindings::spi_device {
+        self.0
+    }
+}
+
+pub struct DriverRegistration {
+    this_module: &'static crate::ThisModule,
+    registered: bool,
+    name: CStr<'static>,
+    probe: Option<SpiMethod>,
+    remove: Option<SpiMethod>,
+    shutdown: Option<SpiMethodVoid>,
+    spi_driver: Option<bindings::spi_driver>,
+}
+
+impl DriverRegistration {
+    fn new(
+        this_module: &'static crate::ThisModule,
+        name: CStr<'static>,
+        probe: Option<SpiMethod>,
+        remove: Option<SpiMethod>,
+        shutdown: Option<SpiMethodVoid>,
+    ) -> Self {
+        DriverRegistration {
+            this_module,
+            name,
+            registered: false,
+            probe,
+            remove,
+            shutdown,
+            spi_driver: None,
+        }
+    }
+
+    // FIXME: Add documentation
+    pub fn new_pinned(
+        this_module: &'static crate::ThisModule,
+        name: CStr<'static>,
+        probe: Option<SpiMethod>,
+        remove: Option<SpiMethod>,
+        shutdown: Option<SpiMethodVoid>,
+    ) -> Result<Pin<Box<Self>>> {
+        let mut registration = Pin::from(Box::try_new(Self::new(
+            this_module,
+            name,
+            probe,
+            remove,
+            shutdown,
+        ))?);
+
+        registration.as_mut().register()?;
+
+        Ok(registration)
+    }
+
+    // FIXME: Add documentation
+    pub fn register(self: Pin<&mut Self>) -> Result {
+        let mut spi_driver = bindings::spi_driver::default();
+        spi_driver.driver.name = self.name.as_ptr() as *const c_types::c_char;
+        spi_driver.probe = self.probe;
+        spi_driver.remove = self.remove;
+        spi_driver.shutdown = self.shutdown;
+
+        let this = unsafe { self.get_unchecked_mut() };
+        if this.registered {
+            return Err(Error::EINVAL);
+        }
+
+        this.spi_driver = Some(spi_driver);
+
+        let res = unsafe { bindings::__spi_register_driver(this.this_module.0, &mut spi_driver) };
+
+        match res {
+            0 => {
+                this.registered = true;
+                Ok(())
+            }
+            _ => Err(Error::from_kernel_errno(res)),
+        }
+    }
+}
+
+impl Drop for DriverRegistration {
+    fn drop(&mut self) {
+        unsafe { bindings::driver_unregister(&mut self.spi_driver.unwrap().driver) }
+        // FIXME: No unwrap? But it's safe?
+    }
+}
+
+// FIXME: Fix SAFETY documentation
+
+// SAFETY: The only method is `register()`, which requires a (pinned) mutable `Registration`, so it
+// is safe to pass `&Registration` to multiple threads because it offers no interior mutability.
+unsafe impl Sync for DriverRegistration {}
+
+// SAFETY: The only method is `register()`, which requires a (pinned) mutable `Registration`, so it
+// is safe to pass `&Registration` to multiple threads because it offers no interior mutability.
+unsafe impl Send for DriverRegistration {}
```

Comment:
Send is about moving T to another thread. Only Sync is about moving &T to another thread.

Reply:
Ah yes, we simply copied this documentation bit from other parts of the Kernel. We definitely need to fix this

Reply:
Fixed by 5ad4cee ?